src/runtime_src/core/pcie/driver/linux/xocl/subdev/lapc.c

```diff
@@ -0,0 +1,234 @@
+/*
+ * A GEM style device manager for PCIe based OpenCL accelerators.
+ *
+ * Copyright (C) 2020 Xilinx, Inc. All rights reserved.
+ *
+ * Authors: Chien-Wei Lan <chienwei@xilinx.com>
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include "../xocl_drv.h"
+
+struct xocl_lapc {
+	void __iomem		*base;
+	struct device		*dev;
+	uint64_t		start_paddr;
+	uint64_t		range;
+	struct mutex 		lock;
+	struct debug_ip_data	data;
+};
+
+static int lapc_remove(struct platform_device *pdev)
+{
+	struct xocl_lapc *lapc;
+	void *hdl;
+
+	lapc = platform_get_drvdata(pdev);
+	if (!lapc) {
+		xocl_err(&pdev->dev, "driver data is NULL");
+		return -EINVAL;
+	}
+
+	xocl_drvinst_release(lapc, &hdl);
+
+	if (lapc->base)
+		iounmap(lapc->base);
+
+	platform_set_drvdata(pdev, NULL);
+
+	xocl_drvinst_free(hdl);
+
+	return 0;
+}
+
+static int lapc_probe(struct platform_device *pdev)
+{
+	struct xocl_lapc *lapc;
+	struct resource *res;
+	int err = 0;
+
+	lapc = xocl_drvinst_alloc(&pdev->dev, sizeof(struct xocl_lapc));
+	if (!lapc)
+		return -ENOMEM;
+
+	lapc->dev = &pdev->dev;
+
+	memcpy(&lapc->data, XOCL_GET_SUBDEV_PRIV(&pdev->dev), sizeof(struct debug_ip_data));
+
+	platform_set_drvdata(pdev, lapc);
+	mutex_init(&lapc->lock);
+
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	if (!res) {
+		err = -ENOMEM;
+		goto done;
+	}
+		
+
+	xocl_info(&pdev->dev, "IO start: 0x%llx, end: 0x%llx",
+		res->start, res->end);
+
+	lapc->base = ioremap_nocache(res->start, res->end - res->start + 1);
+	if (!lapc->base) {
+		err = -EIO;
+		xocl_err(&pdev->dev, "Map iomem failed");
+		goto done;
+	}
+
+	lapc->start_paddr = res->start;
+	lapc->range = res->end - res->start + 1;
+
+done:
+	if (err) {
+		lapc_remove(pdev);
+		return err;
+	}
+	return 0;
+}
+
+static int lapc_open(struct inode *inode, struct file *file)
+{
+	struct xocl_lapc *lapc = NULL;
+
+	lapc = xocl_drvinst_open_single(inode->i_cdev);
+	if (!lapc)
+		return -ENXIO;
+	file->private_data = lapc;
+	return 0;
+}
+
+static int lapc_close(struct inode *inode, struct file *file)
+{
+	struct xocl_lapc *lapc = file->private_data;
+
+	xocl_drvinst_close(lapc);
+	return 0;
+}
+
+long lapc_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
```

Comment:
Why ioctl been implemented by actually doing nothing?

Reply:
Hey @houlz0507 the actual ioctl calls will be implemented by profile/debug team.

Reply:
We may not even need to implement ioctls for lapc and spc as XDP library does not read/write to LAPC/SPC monitors directly. LAPC/SPC are only read by "xbutil". We can just implement/use sysfs nodes for LAPC, SPC for xbutil.